src/vm/runners/builtin_runner/keccak.rs

```diff
@@ -0,0 +1,610 @@
+use crate::vm::errors::memory_errors::MemoryError;
+use crate::vm::errors::runner_errors::RunnerError;
+use crate::vm::vm_core::VirtualMachine;
+use crate::vm::vm_memory::memory::Memory;
+use crate::vm::vm_memory::memory_segments::MemorySegmentManager;
+use num_bigint::BigInt;
+use num_integer::div_ceil;
+
+#[derive(Debug, Clone)]
+pub struct KeccakBuiltinRunner {
+    ratio: u32,
+    pub base: isize,
+    pub(crate) cells_per_instance: u32,
+    pub(crate) n_input_cells: u32,
+    verified_addresses: Vec<Relocatable>,
+    pub(crate) stop_ptr: Option<usize>,
+    _included: bool,
+    state_rep: Vec<u32>,
+    instances_per_component: u32,
+}
+
+impl KeccakBuiltinRunner {
+    pub(crate) fn new(instance_def: &KeccakInstanceDef, included: bool) -> Self {
+        KeccakBuiltinRunner {
+            base: 0,
+            ratio: instance_def._ratio,
+            n_input_cells: 25,
+            cells_per_instance: 50,
+            stop_ptr: None,
+            verified_addresses: Vec::new(),
+            _included: included,
+            instances_per_component: 1,
+            state_rep: vec![
+                64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64, 64,
+                64, 64, 64, 64,
+            ],
+        }
+    }
+
+    pub fn initialize_segments(
+        &mut self,
+        segments: &mut MemorySegmentManager,
+        memory: &mut Memory,
+    ) {
+        self.base = segments.add(memory).segment_index
+    }
+
+    pub fn initial_stack(&self) -> Vec<MaybeRelocatable> {
+        if self._included {
+            vec![MaybeRelocatable::from((self.base, 0))]
+        } else {
+            vec![]
+        }
+    }
+
+    pub fn base(&self) -> isize {
+        self.base
+    }
+
+    pub fn ratio(&self) -> u32 {
+        self.ratio
+    }
+
+    pub fn add_validation_rule(&self, _memory: &mut Memory) -> Result<(), RunnerError> {
+        Ok(())
+    }
+
+    pub fn deduce_memory_cell(
+        &mut self,
+        address: &Relocatable,
+        memory: &Memory,
+    ) -> Result<Option<MaybeRelocatable>, RunnerError> {
+        let index = address.offset % self.cells_per_instance as usize;
+
+        if index < self.n_input_cells as usize {
+            return Ok(None);
+        }
+        let first_input_addr = address
+            .clone()
+            .sub(index)
+            .map_err(|_| RunnerError::BaseNotFinished)?;
+        if self.verified_addresses.contains(&first_input_addr) {
+            return Ok(None);
+        }
+
+        let mut present_in_memory = vec![];
+
+        for i in 0..self.n_input_cells {
+            present_in_memory.push(memory.data.contains(&vec![Some(MaybeRelocatable::from(
+                first_input_addr.clone() + i as usize,
+            ))]))
+        }
+
+        if present_in_memory.iter().all(|&x| x) {
+            return Ok(None);
+        }
+
+        if let Some((i, bits)) = (0..self.state_rep.len()).zip(self.state_rep.clone()).next() {
```

Comment:
This line is a bit confusing, could it be replaced by `for (i,  bits) in self.state_rep.iter().enumerate()` ?